### PR TITLE
Add caffeine cache

### LINKED ISSUE #64: Support Caffeine
https://github.com/ben-manes/caffeine

### PR DESCRIPTION
**Pull request by mchv**
[Caffeine cache library](https://github.com/ben-manes/caffeine) comes with [good documentation](https://github.com/ben-manes/caffeine/wiki) and [top performances](https://github.com/ben-manes/caffeine/wiki/Benchmarks).
Should fix #64

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -160,6 +171,7 @@ object ScalaCacheBuild extends Build {
     libraryDependencies ++= implProjectDeps
   )
 
+  // Maven froufrou
```
Comment (cb372) [inline project/Build.scala]: Wut?

File: caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import com.github.benmanes.caffeine.cache.{ Cache => CCache, Caffeine }
+
+import scala.concurrent.duration.Duration
+import org.joda.time.DateTime
+import com.typesafe.scalalogging.StrictLogging
+import scala.concurrent.Future
+
+/*
+ * Thin wrapper around Caffeine.
+ * Since Guava is in-memory and non-blocking,
```
Comment (cb372) [inline caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala]: s/Guava/Caffeine in a few places in this file

File: caffeine/src/main/scala/scalacache/caffeine/Entry.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package scalacache.caffeine
+
+import org.joda.time.DateTime
+
+/**
+ * A cache entry with an optional expiry time
+ */
+case class Entry[+A](value: A, expiresAt: Option[DateTime]) {
```
Comment (cb372) [inline caffeine/src/main/scala/scalacache/caffeine/Entry.scala]: Let's move this class to the core project and share it across Guava and Caffeine

**Comment by cb372** [comment]
Travis failed because Caffeine requires Java 8:

```
[info] Exception encountered when attempting to run a suite with class name: scalacache.caffeine.CaffeineCacheSpec *** ABORTED ***
[info]   java.lang.UnsupportedClassVersionError: com/github/benmanes/caffeine/cache/Caffeine : Unsupported major.minor version 52.0
```

Not sure what to do about this. Maybe it's time to say goodbye to Java 7, especially as we'll want to start cross-building for Scala 2.12 soon.

File: caffeine/src/main/scala/scalacache/caffeine/Entry.scala (hunk 3 above)
Comment (cb372) [inline caffeine/src/main/scala/scalacache/caffeine/Entry.scala]: Can also share it with LruMap.

**Comment by cb372** [comment]
Thanks for the PR! It's a shame you have to copy-paste so much code. The implementations of `Cache` for Guava, Caffeine and LruMap are pretty much exactly the same as each other. After this is merged I'll do a bit of refactoring and hopefully reduce the copy-pasta.

File: caffeine/src/main/scala/scalacache/caffeine/Entry.scala (hunk 3 above)
Comment (mchv) [inline caffeine/src/main/scala/scalacache/caffeine/Entry.scala]: Ok

File: caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala (hunk 2 above)
Comment (mchv) [inline caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala]: Ok

File: project/Build.scala (hunk 1 above)
Comment (mchv) [inline project/Build.scala]: it was only an easter egg for your reviewer eagle eyes :smile:

**Comment by mchv** [comment]
@cb372 updated changes following your comments.

**Comment by ben-manes** [comment]
Cool! Thanks for integrating. I'm going to add a few comments as improvement ideas.

File: caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+   * @return the value, if there is one
+   */
+  override def get[V](key: String) = {
+    val entry = Option(underlying.getIfPresent(key).asInstanceOf[Entry[V]])
+    /*
+     Note: we could delete the entry from the cache if it has expired,
+     but that would lead to nasty race conditions in case of concurrent access.
+     We might end up deleting an entry that another thread has just inserted.
+     */
+    val result = entry.flatMap { e =>
```
Comment (ben-manes) [inline caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala]: If expired, you could use the `Map.computeIfPresent()` to eagerly evict atomically. The published `RemovalCause` would be `explicit` though, unless you had some adapting scheme.

File: caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+import scala.concurrent.duration.Duration
+import org.joda.time.DateTime
+import com.typesafe.scalalogging.StrictLogging
+import scala.concurrent.Future
+
+/*
+ * Thin wrapper around Caffeine.
+ * Since Caffeine is in-memory and non-blocking,
+ * all operations are performed synchronously, i.e. ExecutionContext is not needed.
```
Comment (ben-manes) [inline caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala]: Both Guava and Caffeine are blocking during the compute. You could use Caffeine's `AsyncLoadingCache` with [scala-java8-compat](https://github.com/scala/scala-java8-compat) to convert to Scala Futures. However, async is incompatible with some features so you might want that to be an alternative adapter.

**Pull request merged**